File: 임현규/programmers/Lv3/lv3_보석쇼핑_상.py
""" Counter를 매 루프마다 생성하는 것은 매우 비효율적이다. 
편의성을 위해 Counter(a) - Counter(b)와 같은 형태로 연산을 수행하려면 
각각의 객체를 생성해야하는데 시간초과를 할 가능성이 높다.

"""
from collections import Counter


def solution(gems):
    kinds = len(set(gems))
    n = len(gems)
    left = right = 0
    answer = [-1, 1_000_000]
    cache = Counter([gems[left]])
    while left <= right and right < n:
        if len(cache) < kinds:
            right += 1
            if right == n:
                break
            cache[gems[right]] += 1

        elif len(cache) == kinds:
            if answer[1] - answer[0] > right - left:
                answer = [left + 1, right + 1]
            cache[gems[left]] -= 1
            if cache[gems[left]] <= 0:
                del cache[gems[left]]
            left += 1
    return answer


print(solution(["DIA", "RUBY", "RUBY", "DIA", "DIA", "EMERALD", "SAPPHIRE", "DIA"]))

File: 임현규/programmers/Lv3/lv3_보석쇼핑_상.java

import java.util.Arrays;
import java.util.HashSet;
import java.util.List;
import java.util.Map;
import java.util.function.Function;
import java.util.stream.Collectors;

class Solution {

    public int[] solution(String[] gems) {
        int kinds = new HashSet<String>(List.of(gems)).size();
        int left = 0;
        int right = 0;
        int[] answer = new int[] {-1, 100000000};
        Counter cache = new Counter(new String[]{gems[0]});
        while (left <= right && right < gems.length) {
            if (cache.getSize() < kinds) {
                right++;
                if (right >= gems.length) {
                    break;
                }
                cache.add(gems[right]);
            }
            else {
                if (right - left < answer[1] - answer[0]) {
                    answer[0] = left + 1;
                    answer[1] = right + 1;
                }
                cache.sub(gems[left]);
                left++;
            }
        }
        return answer;
    }
}

class Counter {

    private final Map<String, Long> map;

    public Counter(String[] list) {
        map = Arrays.stream(list)
            .collect(Collectors.groupingBy(Function.identity(), Collectors.counting()));
    }

    public int getSize() {
        return map.size();
    }

    public void add(String item) {
        map.put(item, map.getOrDefault(item, 0L) + 1);
    }

    public void sub(String item) {
        map.put(item, map.getOrDefault(item, 0L) - 1);
        if (map.get(item) <= 0) {
            map.remove(item);
        }
    }

    @Override
    public String toString() {
        return "Counter{" +
            "map=" + map +
            '}';
    }
}